File: nxflix-agents-ts/src/models/kanji.ts
import { z } from 'zod';

/**
 * A compound word containing a kanji character.
 */
export const CompoundWord = z.object({
  word: z.string(),
  reading: z.string(),
  meaning: z.string(),
});
export type CompoundWord = z.infer<typeof CompoundWord>;

/**
 * A JLPT N1 kanji item with readings, meanings, and related compounds.
 */
export const KanjiItem = z.object({
  id: z.string(),
  character: z.string().length(1),
  strokeCount: z.number().int().min(1),
  onyomi: z.array(z.string()).default([]),
  kunyomi: z.array(z.string()).default([]),
  meanings: z.array(z.string()),
  radicals: z.array(z.string()).default([]),
  compoundWords: z.array(CompoundWord).default([]),
  mnemonics: z.string().nullish(),
  level: z.string().default('N1'),
  contentType: z.literal('kanji').default('kanji'),
});
export type KanjiItem = z.infer<typeof KanjiItem>;

/**
 * Request to generate a set of kanji items.
 */
export const KanjiGenerateRequest = z.object({
  characters: z.array(z.string()).optional(),
  count: z.number().int().min(1).max(50).default(10),
  topic: z.string().optional(),
  includeCompounds: z.boolean().default(true),
});
export type KanjiGenerateRequest = z.infer<typeof KanjiGenerateRequest>;

File: nxflix-agents-python/src/nxflix_agents/models/kanji.py
"""Kanji model for JLPT N1 study."""

from typing import Literal
from pydantic import BaseModel, Field


class CompoundWord(BaseModel):
    """A compound word containing a kanji character."""

    word: str
    reading: str
    meaning: str


class KanjiItem(BaseModel):
    """A JLPT N1 kanji item with readings, meanings, and related compounds."""

    id: str
    character: str = Field(..., min_length=1, max_length=1)
    stroke_count: int = Field(..., ge=1)
    onyomi: list[str] = []
    kunyomi: list[str] = []
    meanings: list[str]
    radicals: list[str] = []
    compound_words: list[CompoundWord] = []
    mnemonics: str | None = None
    level: str = "N1"
    content_type: Literal["kanji"] = "kanji"


class KanjiGenerateRequest(BaseModel):
    """Request to generate a set of kanji items."""

    characters: list[str] | None = None
    count: int = Field(default=10, ge=1, le=50)
    topic: str | None = None
    include_compounds: bool = True
